recirq/qml_lfe/learn_states_q.py

```diff
@@ -0,0 +1,194 @@
+
+    # Add measurements.
+    for i, qubit in enumerate(all_qubits):
+        ret_circuit += cirq.measure(qubit, key=f"q{i}")
+
+    # Merge single qubit operations, flatten moments and align measurements.
+    cirq.merge_single_qubit_gates_into_phxz(ret_circuit)
+    cirq.DropEmptyMoments().optimize_circuit(circuit=ret_circuit)
+    ret_circuit = run_config.flatten_circuit(ret_circuit)
+    cirq.SynchronizeTerminalMeasurements().optimize_circuit(circuit=ret_circuit)
+
+    # Create randomized flippings. These flippings will contain values of 1,0.
+    # which will turn the X gates on or off.
+    params = _create_randomized_sweeps(pauli, flip_params, num_reps)
+    logging.debug(
+        f"Generated circuit w/ depth {len(ret_circuit)} and {len(params)} sweeps."
+    )
+    return ret_circuit, params
+
+
+def run_and_save(
+    n: int,
+    n_paulis: int,
+    batch_size: int,
+    n_shots: int,
+    save_dir: str,
+    use_engine: bool,
+) -> None:
+    """Run and save bitstring data for I + P experiment w/ twocopy.
+
+    Note: uses qubit layouts native to a Sycamore device (Weber).
+
+    Args:
+        n: Number of system qubits to use (total qubits == 2 * n).
+        n_paulis: Number of pauli circuits to generate.
+        batch_size: The number of circuits to send off for execution in
+            a single payload (Does not affect experimental resluts).
+        n_shots: Number of shots to draw from each circuit.
+        save_dir: str or Path to directory where data is saved.
+        use_engine: Whether or not to make use of quantum engine and the
+            weber processor. Note this requires the GOOGLE_CLOUD_PROJECT
+            environment variable to be set, along with the required cloud
+            permissions.
+    """
+    logging.info("Beginning quantum-enhanced circuit generation.")
+    system_pairs = run_config.qubit_pairs()
+    system_pairs = system_pairs[:n]
+
+    logging.info("Generating pauli strings.")
+    paulis = np.array(["X", "Y", "Z", "I"])
+    pauli_strings = np.random.choice(a=paulis, size=(n_paulis, n), replace=True)
+
+    for pauli in pauli_strings:
+        logging.info(f"Processing pauli: {pauli}")
+        circuit, sweeps = build_circuit(system_pairs, pauli, n_shots)
+
+        all_results = []
+        for b in range(0, n_shots, batch_size):
+            results = run_config.execute_sweep(
+                circuit, sweeps[b : b + batch_size], use_engine
+            )
+
+            batch_results = []
+            for j, single_circuit_samples in enumerate(results):
+                qubit_order = [f"q{i}" for i in range(2 * n)]
+                out0 = single_circuit_samples.data[qubit_order].to_numpy()
+                batch_results.append(np.squeeze(out0))
+
+            batch_results = np.array(batch_results)
+            all_results.append(batch_results)
+
+        all_results = np.concatenate(all_results)
+        name0 = "Q-size-{}-pauli-{}".format(n, "".join(t for t in pauli))
```

Comment:
nit:  why name0?  Would filename be a better name?

Reply:
Changed.